### ISSUE TITLE
number_separator rule changed 

### ISSUE THREAD

**Issue by Nef10** [realm/SwiftLint#4637]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

With https://github.com/realm/SwiftLint/pull/4583 the rule change - I think in an unindented way:

| Code | SwiftLint 0.49.0 | SwiftLint 0.50.0 | SwiftLint 0.50.1 |
| ------------- | ------------- | ------------- | ------------- |
| `Date(timeIntervalSince1970: 1497078000)` | error | error | ok |
| `Date(timeIntervalSince1970: 1_497_078_000)` | ok | ok | error |
| `#colorLiteral(red: 0.3543982506, green: 1.0, blue: 1.0, alpha: 1)` | ok | error | ok

I think @jpsim thought that it broke for all decimals (e.g. like line one behaved like line 3) - however the refactoring from 0.50.0 actually only seems to impact the parsing of the colorLiteral.

Additionally, I think the error on `Date(timeIntervalSince1970: 1_497_078_000)`  could be improved, it currently says `Underscores should be used as thousand separator in large decimal numbers. (number_separator)` - would be nice if that is negated if we enforce no `_` for decimals.

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? see above
* Installation method used (Homebrew, CocoaPods, building from source, etc)? portable from GitHub releases
* Paste your configuration file:

```yml
opt_in_rules:
  - number_separator
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? no
* Which Xcode version are you using (check `xcodebuild -version`)? 14.1
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules` 
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`. see above

**Comment by SimplyDanny** [comment]
Checking with 0.50.1 without any configuration I get exactly the same results as in 0.49.0:
| Code | SwiftLint 0.50.1 |
| ------------- | ------------- |
| `Date(timeIntervalSince1970: 1497078000)` | error |
| `Date(timeIntervalSince1970: 1_497_078_000)` | ok |
| `#colorLiteral(red: 0.3543982506, green: 1.0, blue: 1.0, alpha: 1)` | ok |

Can you please confirm that 0.50.1 is used and no other options are specified in the configuration, @Nef10? From what I see it works as expected now.

**Comment by Nef10** [comment]
Sorry, @SimplyDanny seems like I copy pasted the wrong example 🤦 
It should be `Decimal(10.000_00)` and `Decimal(10.00000)` respectively - the one I pasted does't have a decimal point.

Full test:
```
➜ ~/Downloads/portable_swiftlint/swiftlint --version                
0.49.0

➜ echo "Decimal(10.000_00)" | ~/Downloads/portable_swiftlint/swiftlint lint --no-cache --use-stdin --enable-all-rules         
Done linting! Found 0 violations, 0 serious in 1 file.

➜  swiftlint --version
0.50.0

➜ echo "Decimal(10.000_00)" | swiftlint lint --no-cache --use-stdin --enable-all-rules
The `anyobject_protocol` rule is now deprecated and will be completely removed in a future release.
<nopath>:1:9: warning: No Magic Numbers Violation: Magic numbers should be replaced by named constants. (no_magic_numbers)
Done linting! Found 1 violation, 0 serious in 1 file.

➜ ~/Downloads/portable_swiftlint-2/swiftlint --version
0.50.1

➜ echo "Decimal(10.000_00)" | ~/Downloads/portable_swiftlint-2/swiftlint lint --no-cache --use-stdin --enable-all-rules
The `anyobject_protocol` rule is now deprecated and will be completely removed in a future release.
<nopath>:1:9: warning: No Magic Numbers Violation: Magic numbers should be replaced by named constants. (no_magic_numbers)
<nopath>:1:9: warning: Number Separator Violation: Underscores should be used as thousand separator in large decimal numbers. (number_separator)
Done linting! Found 2 violations, 0 serious in 1 file.
```

**Comment by SimplyDanny** [comment]
Now I get it. My sentence

> The separator can be enforced or the rule can be deactivated for decimals. In the inactive state, it will not complain in case _ is still in use.

from above is wrong in that regard. The rule indeed complains and this is exactly the potential bug.

I think this issue boils down to: Should `let i = 1_23` trigger by default and should `let d = 1.234_567` trigger if `minimum_fraction_length` is not set in the configuration? In other words, is the rule expected to trigger also in case a `_` would not be required by it even though the compiler is happy with it?

**Comment by Nef10** [comment]
Hey,
I think there are still two problems. One is the one you mention - should the rule enforce there are no additional `_`. But the other one is, that I still think it should revert back to the 0.49.0 behaviour that it enforces underscores for decimals by default. 

In regards to the colorLiteral issue, maybe we need to add a flag to allow disabling it.

**Comment by SimplyDanny** [comment]
I've created a PR for the first part. With the change, the rule would also trigger on misplaced separators.

About the changed behavior: Actually, `#colorLiteral` also works with numbers containing separators. So I'm not quite sure why it led to the behavioral change. Still, it is how it is and I don't think we should change the behavior again. Ultimately, this is all about the default configuration only. The rule can just be configured in a different way to let it behave like in 0.49.0.

**Comment by NikKovIos** [comment]
Just add - number separator works incorrect. There are automatic changes that it does.

<img width="828" alt="image" src="https://user-images.githubusercontent.com/16647690/221935116-094c9d1a-0b4c-4bb8-a549-05c3f523eceb.png">

**Comment by jpsim** [comment]
@NikKovIos this is with `main`?

**Comment by SimplyDanny** [comment]
> @jpsim what do you mean `main`?

Which version of SwiftLint are you referring to?

**Comment by NikKovIos** [comment]
@SimplyDanny `pod 'SwiftLint', '= 0.50.3'`

**Comment by SimplyDanny** [comment]
So could you please try with `0.51.0-rc.2`?

**Comment by NikKovIos** [comment]
@SimplyDanny yes, help me how to set it in Podfile please? `pod 'SwiftLint', '= 0.51.0-rc.2'` not works.

**Comment by jpsim** [comment]
The first line in the release notes says this: https://github.com/realm/SwiftLint/releases/tag/0.51.0-rc.2

> This is a prerelease version. It won't be published to Homebrew, CocoaPods or bzlmod. But there are many other ways to install:

That being said, if there have been no issues reported with the release candidate we could cut an official release.

**Comment by NikKovIos** [comment]
@jpsim thx, however it's too long to build and replace it for me. I think you can test it by yourself with code:

```
    "lon": 129.811_15,
    "lat": 49.813_299_999_999_998,
    checkMarkPath.move(to: CGPoint(x: frame.height / 3.1578, y: frame.height / 2))
    checkMarkPath.addLine(to: CGPoint(x: frame.height / 2.0618, y: frame.height / 1.57894))
    checkMarkPath.addLine(to: CGPoint(x: frame.height / 1.395_3, y: frame.height / 2.727_2))
```

It's the only strings where rule has incorrect work

**Comment by SimplyDanny** [comment]
I don't get any violations and thus any auto-fixes with this code. Therefore, I suspect your issues to be revolved in the upcoming release.

**Comment by NikKovIos** [comment]
@SimplyDanny thx, waiting for release)
